rust/regex-rs/src/lib.rs

```
@@ -0,0 +1,285 @@
+    pub fn matches(
+        &self,
+        input: &str,
+        max_matches: usize,
+        flags: MatchFlags,
+    ) -> Result<Vec<Match>, String> {
+        let input =
+            CString::new(input).map_err(|_| String::from("Input string contains NUL byte"))?;
+
+        let mut pmatch: Vec<regmatch_t>;
+
+        let errcode = unsafe {
+            pmatch = vec![mem::zeroed(); max_matches];
+
+            regexec(
+                &self.regex,
+                input.into_raw(),
+                max_matches as libc::size_t,
+                pmatch.as_mut_ptr(),
+                flags.bits(),
+            )
+        };
+
+        match errcode {
+            0 => {
+                // Success. Let's copy results
+                let mut matches: Vec<Match> = Vec::new();
+
+                for m in pmatch {
+                    if m.rm_so < 0 || m.rm_eo < 0 {
+                        // Since `max_matches` can be bigger than the number of parenthesized
+                        // blocks in the regex, it's possible that some of the `pmatch` values are
+                        // empty. We exit the loop after detecting first such value.
+                        break;
+                    }
+
+                    // It's safe to cast i32 to usize here:
+                    // - we already checked that the values aren't negative
+                    // - usize's upper bound is higher than i32's
+                    matches.push(Match {
+                        start_pos: m.rm_so as usize,
+                        end_pos: m.rm_eo as usize,
+                    });
+                }
+
+                Ok(matches)
+            }
+
+            libc::REG_NOMATCH => {
+                // Matching went okay, but nothing found
+                Ok(Vec::new())
+            }
+
+            // POSIX only specifies two return codes for regexec(), but implementations are free to
+            // extend that.
+            _ => Err(format!("({}) match failure", errcode)),
```

Might be nice to mention in the error message that the number is an error code (and maybe nice to mention regexec)

---

Just to be clear, I don't know much about error handling in Rust so feel free to ignore.
 Especially if it is normal to have messages without context in the string.

---

> Might be nice to …

Agreed. I also think I should've marked those for translation. Will fix.

> Especially if it is normal to have messages without context in the string.

What context do you mean? I don't think this error string should mention that it comes from `matches()` method, regardless of the language. The caller knows what they're calling, and if they want to propagate the error further, it's their job to add context: they might want to mention `matches()`, or maybe they'll mention the function they're calling `matches()` from, or maybe something else. Basically, it's easy for them to add to the string, but hard to remove from it.

---

I rephrased the error message, marked it for translation, and ran `make extract`. Please take a look.